### PR TITLE
[PWA-2065] Add reCaptcha in checkout

### PR DESCRIPTION
**Pull request by jeremycharron**
## Description

Add reCaptcha in checkout on the Place Order and Braintree payment form

- Also fixes issue where reCaptcha badge was still rendering on form even if reCaptcha config was turned off in Admin.
- Also fixes issue where first reCaptcha was not initialized properly if another reCaptcha was initialized at the same time.
- Also fixes issue where "Place Order" button was enabling/disabling after click. This was also occurring on production but so quickly that it was impossible to see with human eyes.

**ReCaptcha must be activated in the backend for Place Order and Braintree payment form**

Back-end pull-request : https://github.com/magento-commerce/magento2-pwa/pull/24

A fully functional test environment is available here : https://jeremycharron-o4saqja-4txwfa4z2sor6.eu-4.magentosite.cloud/
Please get in touch with me if you need an Admin account.

## Related Issue

Closes PWA-2065

### Verification Stakeholders

Project maintainers

## Verification Steps

**ReCaptcha must be activated in the backend for Place Order and Braintree payment form**

#### Test scenario(s) for direct fix/feature

1. With reCaptcha threshold to 1.0, input credit card and click "Review order", then validate that you have a "reCaptcha validation failed error"
2. With reCaptcha threshold to 0.0, input credit card and click "Review order", then validate that credit card was succesfully added
3. With reCaptcha threshold to 1.0, click on "Place order" then validate that you have a "reCaptcha validation failed error"
4. With reCaptcha threshold to 0.0, click on "Place order" then validate that order was successful
5. Make sure that X-Captcha token is passed in the header
6. Test new Cypress test `googleReCaptcha/verifyCheckoutReCaptcha.spec.js`
7. Make sure that multiple reCaptchas in checkout are working (placeOrder + braintree + signIn)

#### Test scenario(s) for any existing impacted features/areas

1. Run Jest tests
2. Run Cypress test `googleReCaptcha/verifyForgetPasswordReCaptcha.spec.js` (was renamed)
3. Test all other reCaptcha (forget password, create account, login) to make sure they still work

#### Test scenario(s) for any Magento Backend Supported Configurations

N/A

#### Is Browser/Device testing needed?

Yes (Chrome, Safari, Firefox, Edge) + mobile

#### Any ad-hoc/edge case scenarios that need to be considered?

No

## Breaking Changes (if any)

No breaking change

## Checklist

-   I have added tests to cover my changes, if necessary.
-   I have added translations for new strings, if necessary.
-   I have updated the documentation accordingly, if necessary.

### CODE REVIEW

**Review by mikhaelbois — APPROVED** [review]
LGTM ✅

**Comment by fooman** [comment]
Has any consideration been given to have all recaptcha functionality in a separate optional package? What if someone doesn’t want to use recaptcha or wants to use a different service?

**Comment by jeremycharron** [comment]
> Has any consideration been given to have all recaptcha functionality in a separate optional package? What if someone doesn’t want to use recaptcha or wants to use a different service?

ReCaptcha must be activated in the backend (with keys generated from Google Recaptcha dashboard) in order to work.

If ReCaptcha is disabled in the Admin, the badge will not be showed and regular flow will occur. This is the same behavior in Luma.

I believe this code wouldn't constraint developers that would want to implement an alternative service.

**Comment by fooman** [comment]
Thanks for confirming. Will flag here though that the current approach taken by the PWA Studio team of adding features increases the burden on existing implementations to keep up. In this case we will now have to review all these recaptcha code additions in the code even though we do not have immediate plans to use it.

I am strongly hoping that venia-ui can be kept as small and light weight as possible and every feature gets implemented as a package. If a straight forward implementation via a package is not possible it highlights a need for better extensibility in venia-ui.

**Comment by dpatil-magento** [comment]
run cypress

**Comment by dpatil-magento** [comment]
@jeremycharron When recaptcha for place order is enabled > For a fraction of second place order button gets enabled in-between a transaction, please take a look.

<a href="https://gyazo.com/18dda7afa496f4bbeb956b16457874a6">Image from Gyazo</a>

**Comment by jeremycharron** [comment]
@dpatil-magento I updated PR to fix issue 1. The button should no longer flash between disabled-enable state after you click.

Please note that this issue was also occurring on develop, but so quickly that it was not visible with human eyes. I was able to replicate with breakpoints. It made visible here because my code adds an `await` for reCaptcha.

I'm currently working on issue 2 which is backend.

**Comment by dpatil-magento** [comment]
run cypress

**Comment by jeremycharron** [comment]
@dpatil-magento I updated [backend pull-request](https://github.com/magento-commerce/magento2-pwa/pull/24) to fix issue 2. I'm not able to replicate issue 3.

File: venia-integration-tests/src/fixtures/googleReCaptchaApi/reCaptchaV3ConfigInline.json
Comment (dpatil-magento) [inline venia-integration-tests/src/fixtures/googleReCaptchaApi/reCaptchaV3ConfigInline.json:10]: same here as well. missing `,` causing parsing error.

File: venia-integration-tests/src/fixtures/googleReCaptchaApi/reCaptchaV3ConfigDefault.json
Comment (dpatil-magento) [inline venia-integration-tests/src/fixtures/googleReCaptchaApi/reCaptchaV3ConfigDefault.json:10]: @jeremycharron missing `,` causing parsing error.

Comment (jeremycharron) [reply venia-integration-tests/src/fixtures/googleReCaptchaApi/reCaptchaV3ConfigDefault.json]: Sorry about that. Fixed !

**Comment by dpatil-magento** [comment]
run cypress

**Review by mikhaelbois — DISMISSED** [review]
LGTM ✅

**Comment by jeremycharron** [comment]
run cypress

**Review by mikhaelbois — APPROVED** [review]

**Comment by jeremycharron** [comment]
run cypress

**Comment by jeremycharron** [comment]
@dpatil-magento Just runned Cypress again. Hopefully, next results should be good.

- Fixed issue where first reCaptcha was not initialized properly if another reCaptcha was initialized at the same time. This is why the Cypress test failed, because Sign In recaptcha was added from develop.
- Fixed another issue where reCaptcha badge was still rendering on form even if reCaptcha config was turned off in Admin.

The changes I made may impact any reCaptcha on the site. I made a general QA to make sure they all still work, but I recommend an additional QA before merging.

**Comment by dpatil-magento** [comment]
QA Approved.

**Comment by fooman** [comment]
Seeing that fixes are seemingly required to the backend which are in a non public repository can some steps be provided as to what these changes are. Currently it sounds likely that all PWA Studio implementations backed by open source will have an issue.

**Comment by jeremycharron** [comment]
@fooman 
In short, two plugins will been added in the following module that was created for the forget password feature. This is part of the Open Source `magento2-pwa` metapackage.  : https://github.com/magento/magento2-pwa/tree/main/ReCaptchaPwa
Code is currently in the internal team fork of `magento2-pwa`, which explains why it's private.

Another metapackage exists for Adobe Commerce, but it's currently empty : https://github.com/magento/magento2-pwa-commerce

**Comment by fooman** [comment]
thanks for the explanation @jeremycharron - will wait for the next public release of magento2-pwa.

**Pull request merged**
